File: lintcode/python/0148_sort_colors.py
#!/usr/bin/python -t

# two pointers

class Solution:
    """
    @param nums: A list of integer which is 0, 1 or 2 
    @return: nothing
    """
    def sortColors(self, nums):
        # write your code here
        n = len(nums)
        if n < 2:
            return
        
        red = 0
        blue = n-1
        i = 0
        
        while i <= blue:
            if nums[i] == 0:
                nums[i], nums[red] = nums[red], nums[i]
                red += 1
                i += 1
            elif nums[i] == 1:
                i += 1
            else:
                nums[i], nums[blue] = nums[blue], nums[i]
                blue -= 1
                
                
        return
    

File: lintcode/golang/0148_sort_colors.go
/**
 * @param nums: A list of integer which is 0, 1 or 2 
 * @return: nothing
 */
func sortColors (nums *[]int)  {
    // write your code here
    n := len(*nums)
    i := 0
    l := 0
    r := n-1
    
    for i <= r {
        if (*nums)[i] == 0 {
            (*nums)[i], (*nums)[l] = (*nums)[l], (*nums)[i]
            i++
            l++
        } else if (*nums)[i] == 1 {
            i++
        } else if (*nums)[i] == 2 {
            (*nums)[i], (*nums)[r] = (*nums)[r], (*nums)[i]
            r--
        }
    }
    
    return
}

